lib/function-promoter.js

```diff
@@ -0,0 +1,67 @@
+const validateArgumentTransformers = require('./argument-transformer-validator');
+const RiffError = require('./riff-error');
+const MappingTransform = require('./mapping-transform');
+const logger = require('util').debuglog('riff');
+
+const withTransformers = (promotedFunction, userFunction) => {
+    const transformers = userFunction['$argumentTransformers'];
+    if (typeof transformers === 'undefined') {
+        return promotedFunction;
+    }
+    validateArgumentTransformers(transformers);
+    const transformerCount = transformers.length;
+    if (transformerCount !== 1) {
+        throw new RiffError(
+            'error-argument-transformer',
+            `Request-reply function must declare exactly 1 argument transformer. Found ${transformerCount}`
+        );
+    }
+    promotedFunction['$argumentTransformers'] = transformers;
+    return promotedFunction;
+};
+
+const withHooks = (promotedFunction, userFunction) => {
+    const initHook = userFunction['$init'];
+    validateHook(initHook, 'init');
+    if (typeof initHook === 'function') {
+        promotedFunction['$init'] = userFunction['$init'];
+    }
+
+    const destroyHook = userFunction['$destroy'];
+    validateHook(destroyHook, 'destroy');
+    if (typeof destroyHook === 'function') {
+        promotedFunction['$destroy'] = userFunction['$destroy'];
+    }
+    return promotedFunction;
+};
+
+const validateHook = (hook, hookName) => {
+    const hookType = typeof hook;
+    if (hookType !== 'function' && hookType !== 'undefined') {
+        throw new RiffError('error-promoting-function', `Request-reply function ${hookName} hook must be a function. Found: ${hookType}`);
+    }
+};
+
+module.exports = (userFunction) => {
+    const interactionModel = userFunction['$interactionModel'] || 'request-reply';
+    if (interactionModel !== 'request-reply') {
+        return userFunction;
+    }
+
+    if (userFunction.length !== 1) {
```

Comment:
attempting to do anything with a function's length is notoriously unreliable and should be avoided.

This is a perfectly legit function that will fail this test:

```js
module.exports = function() { return arguments[0] ** 2; }
```

as would:

```js
module.exports = (x, yIsThisEvenHere) => x ** 2;
```